
if "--use-singularity" in sys.argv:
    worflow_path = workflow.snakefile.split("workflows/full_pipeline.rules")[0]
    workflow.singularity_args += f' -B {worflow_path}:{worflow_path}'

include:
    "resistance.rules"

include:
    "virulence.rules"

include:
    "benchmark.rules"

rule epidemiology:
    input:
        "report/report_epidemiology.html",

rule virulence:
    input:
        "report/report_virulence.html",

rule resistance:
    input:
        "report/report_resistance.html",

rule strain_characterization:
    input:
        "report/report_strain_characterization.html",

rule custom_db:
    input:
        # expand("samples/{sample}/resistance/bwa/walker_resistant_annotated/mutations.vcf", sample=read_naming.keys()),
        # expand("samples/{sample}/resistance/bwa/mykrobe_annotated/mutations.vcf", sample=read_naming.keys()),
        "report/resistance/rgi_benchmark.tsv",
        "report/resistance/mykrobe_benchmark.tsv",
        "report/resistance/bwa/walker_resistant_annotated_benchmark.tsv",
        "report/resistance/tb-profiler/tbprofiler_benchmark.txt",
        # expand("samples/{sample}/resistance/tb-profiler/results/{sample}.results.json", sample=read_naming.keys())


'''
- we set a rule order to be able to extract core SNP data from the complete reference genome data
(without mapping of the linked cgMLST genome which is the same as the reference one)
'''
