

import re
import yaml
import os

PIPELINE_VERSION = "2.4.0"


if "--use-singularity" in sys.argv:
    worflow_path = workflow.snakefile.split("workflows/full_pipeline.rules")[0]
    workflow.singularity_args += f' -B {worflow_path}:{worflow_path}'

    singularity_envs = yaml.safe_load(open(os.path.join(workflow.basedir,  "../envs/singularity/envs.yml"), 'r'))

    
include:
    "resistance.rules"

include:
    "virulence.rules"

include:
    "benchmark.rules"


def get_assembly_list():
    assembly_out =[]
    for reference in reference_genomes:
        if 'assembled_genome' in reference:
            sample = re.match("full_genome_(.*)_assembled_genome", reference).group(1)
            assembly_out.append(f"report/genomes/{sample}.faa")
            assembly_out.append(f"report/genomes/{sample}.fna")
            assembly_out.append(f"report/genomes/{sample}.gbk")
    return assembly_out


rule version:
    output:
        "report/pipeline_version.txt"
    run:
        with open(output[0], 'w') as f:
            f.write(f"Diag_pipeline version {PIPELINE_VERSION}\n")


rule epidemiology:
    input:
        "report/report_epidemiology.html",
        get_assembly_list()


rule virulence:
    input:
        "report/report_virulence.html",
        expand("report/genomes/{sample}.faa", sample=read_naming.keys()),
        expand("report/genomes/{sample}.gbk", sample=read_naming.keys()),
        expand("report/genomes/{sample}.fna", sample=read_naming.keys()),


rule resistance:
    input:
        "report/report_resistance.html",
        expand("report/genomes/{sample}.faa", sample=read_naming.keys()),
        expand("report/genomes/{sample}.gbk", sample=read_naming.keys()),
        expand("report/genomes/{sample}.fna", sample=read_naming.keys()),

rule strain_characterization:
    input:
        "report/report_strain_characterization.html",

rule custom_db:
    input:
        # expand("samples/{sample}/resistance/bwa/walker_resistant_annotated/mutations.vcf", sample=read_naming.keys()),
        # expand("samples/{sample}/resistance/bwa/mykrobe_annotated/mutations.vcf", sample=read_naming.keys()),
        "report/resistance/tb-profiler/tbprofiler_benchmark_FP.out",
        "report/resistance/rgi_benchmark_FP.out",
        "report/resistance/mykrobe_benchmark_FP.out",
        "report/resistance/bwa/walker_resistant_annotated/walker_resistant_annotated_benchmark_FP.out",
        # expand("samples/{sample}/resistance/tb-profiler/results/{sample}.results.json", sample=read_naming.keys())


'''
- we set a rule order to be able to extract core SNP data from the complete reference genome data
(without mapping of the linked cgMLST genome which is the same as the reference one)
'''
